File: 116.populating_next_right_pointers_in_each_node/connect.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if not root:
            return None
        # connect root.left and root.right
        if root.left:
            root.left.next = root.right
        if root.next and root.next.left:
            root.right.next = root.next.left
        self.connect(root.left)
        self.connect(root.right)

File: 116.populating_next_right_pointers_in_each_node/connect.js
/**
 * Definition for binary tree with next pointer.
 * function TreeLinkNode(val) {
 *     this.val = val;
 *     this.left = this.right = this.next = null;
 * }
 */

/**
 * @param {TreeLinkNode} root
 * @return {void} Do not return anything, modify tree in-place instead.
 */
var connect = function(root) {
    // connect all the children
    if(root === null) return;

    if (root.left) {
        root.left.next = root.right;
    }
    // connect nodes between subtrees
    if (root.next && root.next.right){
        root.right.next = root.next.left;
    }

    connect(root.left);
    connect(root.right);

};